From 8ba8bf36997ba5726098cedad261663129f4d806 Mon Sep 17 00:00:00 2001
From: bins <bins@kp-test>
Date: Wed, 7 Oct 2020 04:28:44 -0400
Subject: [PATCH] kernel struct assignment instrument

---
 kernel/kcov.c | 62 +++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 62 insertions(+)

diff --git a/kernel/kcov.c b/kernel/kcov.c
index 6afae0bcbac4..b6a40fc40a94 100644
--- a/kernel/kcov.c
+++ b/kernel/kcov.c
@@ -322,6 +322,68 @@ void notrace __sanitizer_cov_trace_switch(u64 val, u64 *cases)
 EXPORT_SYMBOL(__sanitizer_cov_trace_switch);
 #endif /* ifdef CONFIG_KCOV_ENABLE_COMPARISONS */
 
+static void notrace write_srt_data(u64 id, u64 val, u64 ip)
+{
+	struct task_struct *t;
+	u64 *area;
+	u64 count, start_index, end_pos, max_pos;
+
+	t = current;
+
+	if (!check_kcov_mode(KCOV_MODE_TRACE_PC, t))
+		return;
+
+	ip = canonicalize_ip(ip);
+
+	area = (u64 *)t->kcov_area;
+	max_pos = t->kcov_size * sizeof(unsigned long);
+
+	count = READ_ONCE(area[0]);
+
+	/* Every record is KCOV_WORDS_PER_STRU 64-bit words. */
+	start_index = 1 + count;
+	end_pos = (start_index + 3) * sizeof(u64);
+	if (likely(end_pos <= max_pos)) {
+		area[start_index] = id;
+		area[start_index + 1] = val;
+		area[start_index + 2] = ip;
+		WRITE_ONCE(area[0], count + 3);
+	}
+}
+
+void notrace __sanitizer_cov_trace_srt1(u64 id, u8 val)
+{
+	u64 id_64 = ((u64)0xefe1 << 48) | (u64)id;
+	u64 val_64 = (u64)val;
+	write_srt_data(id_64, val_64, _RET_IP_);
+}
+EXPORT_SYMBOL(__sanitizer_cov_trace_srt1);
+
+void notrace __sanitizer_cov_trace_srt2(u64 id, u16 val)
+{
+	u64 id_64 = ((u64)0xefe2 << 48) | (u64)id;
+	u64 val_64 = (u64)val;
+	write_srt_data(id_64, val_64, _RET_IP_);
+}
+EXPORT_SYMBOL(__sanitizer_cov_trace_srt2);
+
+void notrace __sanitizer_cov_trace_srt4(u64 id, u32 val)
+{
+	u64 id_64 = ((u64)0xefe4 << 48) | (u64)id;
+	u64 val_64 = (u64)val;
+	write_srt_data(id_64, val_64, _RET_IP_);
+}
+EXPORT_SYMBOL(__sanitizer_cov_trace_srt4);
+
+void notrace __sanitizer_cov_trace_srt8(u64 id, u64 val)
+{
+	u64 id_64 = ((u64)0xefe8 << 48) | (u64)id;
+	u64 val_64 = (u64)val;
+	write_srt_data(id_64, val_64, _RET_IP_);
+}
+EXPORT_SYMBOL(__sanitizer_cov_trace_srt8);
+
+
 static void kcov_start(struct task_struct *t, struct kcov *kcov,
 			unsigned int size, void *area, enum kcov_mode mode,
 			int sequence)
-- 
2.20.1

